src/main/java/org/hydev/mcpm/client/injector/PluginLoader.java

```diff
@@ -1,45 +1,135 @@
+import static org.hydev.mcpm.utils.ReflectionUtils.setPrivateField;
 
 /**
  * TODO: Write a description for this class!
  *
  * @author Azalea (https://github.com/hykilpikonna)
  * @since 2022-09-27
  */
-public class PluginLoader
+public class PluginLoader implements LoadBoundary, UnloadBoundary, ReloadBoundary
 {
-    /**
-     * Dynamically load a local plugin through JVM reflections and classloader hacks
-     *
-     * @param jar Local jar file path
-     */
+    @Override
+    public void loadPlugin(String name) throws PluginNotFoundException
+    {
+        // 1. Find plugin file by name
+        var nf = new PluginNotFoundException(name);
+        var dir = new File("plugins");
+        if (!dir.isDirectory()) throw nf;
+        var file = Arrays.stream(Optional.ofNullable(dir.listFiles()).orElseThrow(() -> nf))
+            .filter(f -> f.getName().endsWith(".jar"))
+            .filter(f -> {
+                try (var jf = new PluginJarFile(f))
+                {
+                    return jf.readPluginYaml().getName().equalsIgnoreCase(name);
+                }
+                catch (IOException ignored) { return false; }
+            }).findFirst().orElseThrow(() -> nf);
+
+        loadPlugin(file);
+    }
+
+    @Override
     public void loadPlugin(File jar)
     {
-        // TODO: Implement this
-        throw new UnsupportedOperationException("TODO");
+        // 2. Load plugin
+        var pm = Bukkit.getPluginManager();
+        try
+        {
+            var plugin = pm.loadPlugin(jar);
+            assert plugin != null;
+
+            // 3. Call onLoad()
+            plugin.onLoad();
+            pm.enablePlugin(plugin);
+        }
+        catch (InvalidPluginException | InvalidDescriptionException e)
+        {
+            // These are errors indicating that the plugin we're trying to load is badly formatted.
+            // There are nothing we can do.
+            e.printStackTrace();
+        }
     }
 
-    /**
-     * Dynamically unload a local plugin through JVM reflections and classloader hacks
-     *
-     * @param name Loaded plugin name
-     */
-    public void unloadPlugin(String name)
+    @Override
+    public void unloadPlugin(String name) throws PluginNotFoundException
     {
-        // TODO: Implement this
-        throw new UnsupportedOperationException("TODO");
+        var pm = Bukkit.getPluginManager();
+
+        // 1. Find plugin by name
+        var plugin = Arrays.stream(pm.getPlugins()).filter(p -> p.getName().equalsIgnoreCase(name)).findFirst()
+            .orElseThrow(() -> new PluginNotFoundException(name));
+
+        // 2. Unload plugin
+        pm.disablePlugin(plugin);
+
+        // 3. Unregister listeners
+        getPrivateField(pm, "listeners", new TypeToken<Map<Event, SortedSet<RegisteredListener>>>(){})
+            .ifPresent(listeners -> listeners.values().forEach(set -> set.removeIf(l -> l.getPlugin() == plugin)));
+
```

Comment:
Would it be possible to fork spigot to expose this? I'm not a fan of these kinds of sneaky maneuvers. If you bump your spigot version each one of these statements might suddenly break.

Reply:
But we are just a package manager, we kind of don't have the authority to ask Spigot to support us, or ask any server admin to use our own fork of Spigot. There are already many forks of Spigot available, like PaperMC, Tuinity, Glowstone, Purpur (which I personally use), and they all come with different levels of tweaks and optimizations. We can't ask any admin to use our unoptimized fork.

Reply:
We're not asking spigot to support us. By forking spigot we're opting into fixing changes that might affect our success at compile/merge time.

Are we trying to write code that supports all the forks that you've mentioned 😅? I feel like each fork could have different optimizations/layouts/values for the classes we're trying to manipulate.

Reply:
They actually don't. Their optimizations are mostly on in-game logic (it probably makes sense because plugin loading is normally only done once when the server starts, so there is not much need for optimizations)

In fact, I tested this on Purpur, and it works even when Purpur is a fork of Tuinity which is a fork of Paper which is a fork of Spigot which is a fork of CraftBukkit.

Reply:
We can keep using reflection then :eyes:

I think optimizing how plugins are loaded seems like a reasonable thing to expect a fork to do, whether it's for performance (I want to precompute these things to save time later!) or for utility of the server operator (let's make it easy for the operator to change things). I think you've made some really good points though, I think I'm just suggesting doing a lot of work for something that ultimately makes our tool *less* useful.